File: app/src/main/java/com/example/data/remote/XBookmarksApi.kt
package com.example.data.remote

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.Path
import retrofit2.http.Query

@JsonClass(generateAdapter = true)
data class TweetAttachmentsDto(
    @Json(name = "media_keys") val mediaKeys: List<String>? = null
)

/** Long-form tweet body (>280 chars). Prefer this over `text` when present. */
@JsonClass(generateAdapter = true)
data class NoteTweetDto(
    @Json(name = "text") val text: String? = null
)

/** A single expanded URL entity (t.co → real link). */
@JsonClass(generateAdapter = true)
data class UrlEntityDto(
    @Json(name = "url") val url: String? = null,
    @Json(name = "expanded_url") val expandedUrl: String? = null,
    @Json(name = "display_url") val displayUrl: String? = null
)

@JsonClass(generateAdapter = true)
data class TweetEntitiesDto(
    @Json(name = "urls") val urls: List<UrlEntityDto>? = null
)

@JsonClass(generateAdapter = true)
data class BookmarkDto(
    @Json(name = "id") val id: String,
    @Json(name = "text") val text: String,
    @Json(name = "created_at") val createdAt: String? = null,
    @Json(name = "author_id") val authorId: String? = null,
    @Json(name = "note_tweet") val noteTweet: NoteTweetDto? = null,
    @Json(name = "entities") val entities: TweetEntitiesDto? = null,
    @Json(name = "attachments") val attachments: TweetAttachmentsDto? = null
)

@JsonClass(generateAdapter = true)
data class MediaDto(
    @Json(name = "media_key") val mediaKey: String,
    @Json(name = "type") val type: String? = null,
    @Json(name = "url") val url: String? = null,
    @Json(name = "preview_image_url") val previewImageUrl: String? = null,
    @Json(name = "alt_text") val altText: String? = null
)

/** Expanded author from `includes.users`, joined by author_id. */
@JsonClass(generateAdapter = true)
data class UserDto(
    @Json(name = "id") val id: String,
    @Json(name = "name") val name: String? = null,
    @Json(name = "username") val username: String? = null
)

@JsonClass(generateAdapter = true)
data class BookmarksIncludesDto(
    @Json(name = "media") val media: List<MediaDto>? = null,
    @Json(name = "users") val users: List<UserDto>? = null
)

@JsonClass(generateAdapter = true)
data class BookmarksMetaDto(
    @Json(name = "next_token") val nextToken: String? = null
)

@JsonClass(generateAdapter = true)
data class BookmarksResponse(
    @Json(name = "data") val data: List<BookmarkDto>? = null,
    @Json(name = "includes") val includes: BookmarksIncludesDto? = null,
    @Json(name = "meta") val meta: BookmarksMetaDto? = null
)

interface XBookmarksApi {
    @GET("2/users/{userId}/bookmarks")
    suspend fun getBookmarks(
        @Header("Authorization") authHeader: String,
        @Path("userId") userId: String,
        @Query("max_results") maxResults: Int = 100,
        @Query("pagination_token") paginationToken: String? = null,
        @Query("tweet.fields") tweetFields: String = "created_at,attachments,author_id,note_tweet,entities",
        @Query("expansions") expansions: String = "attachments.media_keys,author_id",
        @Query("media.fields") mediaFields: String = "url,preview_image_url,type,alt_text",
        @Query("user.fields") userFields: String = "name,username"
    ): BookmarksResponse
}

File: ios/Curio/Curio/Networking/XBookmarksApi.swift
import Foundation

/// X v2 bookmarks endpoint. Ports `interface XBookmarksApi` in `data/remote/XBookmarksApi.kt`.
///
/// Paginated GET to `https://api.twitter.com/2/users/{userId}/bookmarks` with the exact fixed query
/// expansions/fields the Retrofit `@Query` defaults carried. The `Authorization` header is passed
/// per-call (CONVENTIONS Networking cross-cutting). On non-2xx (e.g. 401 / 429) it throws
/// ``APIError`` so the repository can refresh tokens / read the rate-limit reset header.
protocol XBookmarksApi: Sendable {
    func getBookmarks(
        authHeader: String,
        userId: String,
        maxResults: Int,
        paginationToken: String?
    ) async throws -> BookmarksResponse
}

extension XBookmarksApi {
    /// Default-argument bridge mirroring the Kotlin `maxResults = 100`, `paginationToken = null`.
    func getBookmarks(authHeader: String, userId: String) async throws -> BookmarksResponse {
        try await getBookmarks(authHeader: authHeader, userId: userId, maxResults: 100, paginationToken: nil)
    }
}

/// Concrete `URLSession`-backed bookmarks client. `actor` per CONVENTIONS §5.
actor XBookmarksApiClient: XBookmarksApi {

    /// Base host (exact — CONVENTIONS Networking cross-cutting "authorize host `twitter.com` exact").
    static let baseURL = URL(string: "https://api.twitter.com/")!

    // Fixed v2 query parameters — byte-identical to the Retrofit `@Query` defaults.
    private static let tweetFields = "created_at,attachments,author_id,note_tweet,entities"
    private static let expansions = "attachments.media_keys,author_id"
    private static let mediaFields = "url,preview_image_url,type,alt_text"
    private static let userFields = "name,username"

    private let http: HTTPClient
    private let baseURL: URL

    init(http: HTTPClient, baseURL: URL = XBookmarksApiClient.baseURL) {
        self.http = http
        self.baseURL = baseURL
    }

    func getBookmarks(
        authHeader: String,
        userId: String,
        maxResults: Int = 100,
        paginationToken: String? = nil
    ) async throws -> BookmarksResponse {
        let base = baseURL.appendingPathComponent("2/users/\(userId)/bookmarks")
        var components = URLComponents(url: base, resolvingAgainstBaseURL: false)!
        var items: [URLQueryItem] = [
            URLQueryItem(name: "max_results", value: String(maxResults))
        ]
        if let token = paginationToken {
            items.append(URLQueryItem(name: "pagination_token", value: token))
        }
        items.append(contentsOf: [
            URLQueryItem(name: "tweet.fields", value: Self.tweetFields),
            URLQueryItem(name: "expansions", value: Self.expansions),
            URLQueryItem(name: "media.fields", value: Self.mediaFields),
            URLQueryItem(name: "user.fields", value: Self.userFields)
        ])
        components.queryItems = items

        let request = HTTPClient.jsonRequest(
            url: components.url!,
            method: "GET",
            authorization: authHeader
        )
        return try await http.send(BookmarksResponse.self, request: request, session: .primary)
    }
}
